02_knight_and_princess/robert_gaisin/main.cpp

```diff
@@ -64,33 +69,40 @@ class Graph {
   VertexId next_vertex_id() { return num_of_vrt_++; }
   EdgeId next_edge_id() { return num_of_edg_++; }
 
-  bool connection(const VertexId& begin, const VertexId& end);
+  bool is_vertex(const VertexId& vertex_id) const;
+  bool is_connected(const VertexId& begin, const VertexId& end) const;
   vector<Vertex> vertices_;
   vector<Edge> edges_;
 };
 
-bool Graph::connection(const VertexId& begin, const VertexId& end) {
-  bool ret = false;
-  for (const EdgeId& edge_num : vertices_[begin].edge_ids) {
-    if ((ret = edges_[edge_num].begin == end || edges_[edge_num].end == end)) {
-      break;
+bool Graph::is_vertex(const VertexId& vertex_id) const {
+  for (const auto& vertex : vertices_) {
+    if (vertex_id == vertex.id)
+      return true;
+  }
+  return false;
+}
+
+bool Graph::is_connected(const VertexId& begin, const VertexId& end) const {
+  for (const EdgeId& edge_num : vertices_[begin].edge_ids()) {
+    if (edges_[edge_num].begin == end || edges_[edge_num].end == end) {
+      return false;
     }
   }
-  return ret;
+  return true;
 }
 
 void Graph::add_vertex() {
-  Vertex vertex = {next_vertex_id()};
-  vertices_.push_back(vertex);
+  vertices_.emplace_back(next_vertex_id());
 }
 
 void Graph::add_edge(const VertexId& begin, const VertexId& end) {
-  if ((begin < num_of_vrt_) && (end < num_of_vrt_) && !connection(begin, end)) {
-    Edge edge = {next_edge_id(), begin, end};
-    vertices_[begin].edge_ids.push_back(edge.id);
-    vertices_[end].edge_ids.push_back(edge.id);
-    edges_.push_back(edge);
-  }
+  if (!is_vertex(begin) || !is_vertex(end) || !is_connected(begin, end))
+    return;
```

Comment:
Тут есть сразу несколько проблем с данным кодом:
1. Ты не обрабатываешь ошибку, а скрываешь её. Просто вызывая `return` ты никогда не узнаешь что произошла ошибка, так делать нельзя, это только усложнит дебаг.
2. В данном случае, вообще нет нужды обрабатывать ошибку, так как у нас нет инпута от пользователя из сети, а значит мы не можем ожидать ошибку. Единственное, мы должны проверить самих себя на адекватность и на то, что не был допущен баг. Делается это через `assert`. Они работают только в `debug` билдах, а в продакшене отключаются не влияя на эффективность конечного кода.
3. Ты перепутал `!is_connected` и `is_connected`, тут мы хотим проверить на то, что они не связаны, поэтому нужно было проверять на `is_connected`.
```suggestion
  assert(has_vertex(begin) && "Vertex doesn't exist");
  assert(has_vertex(end) && "Vertex doesn't exist");
  assert(!is_connected(begin, end) && "Vertices already connected");
```